openmmtools/multistate/multistateanalyzer.py

```diff
@@ -2009,28 +2021,35 @@ def _get_equilibration_data(self, energies=None, neighborhoods=None, replica_sta
-        t0 = 1 # discard minimization frame
-        try:
-            iteration = len(u_n)
-            data = self._reporter.read_online_analysis_data(None, 't0')
-            t0 = max(t0, int(data['t0'][0]))
-            logger.debug('t0 found; using initial t0 = {} instead of 1'.format(t0))
-        except Exception as e:
-            # No t0 found
-            logger.debug('Could not find t0: {}'.format(e))
-            pass
+        if self._n_equilibration_iterations is not None and self._statistical_inefficiency is not None:
+            n_equilibration = self._n_equilibration_iterations
+            g_t = self._statistical_inefficiency
+            n_effective_max = (self.max_n_iterations - n_equilibration + 1) / g_t
 
-        # Discard equilibration samples.
-        # TODO: if we include u_n[0] (the energy right after minimization) in the equilibration detection,
-        # TODO:         then number_equilibrated is 0. Find a better way than just discarding first frame.
-        i_t, g_i, n_effective_i = multistate.utils.get_equilibration_data_per_sample(u_n[t0:])
-        n_effective_max = n_effective_i.max()
-        i_max = n_effective_i.argmax()
-        n_equilibration = i_t[i_max] + t0 # account for initially discarded frames
-        g_t = g_i[i_max]
+        else:
+            u_n = self.get_effective_energy_timeseries(energies=energies, neighborhoods=neighborhoods, replica_state_indices=replica_state_indices)
+
+            # For SAMS, if there is a second-stage start time, use only the asymptotically optimal data
+            t0 = self._n_equilibration_iterations if self._n_equilibration_iterations is not None else 1 # if self._n_equilibration_iterations was not specified, discard minimization frame
+            try:
+                iteration = len(u_n)
+                data = self._reporter.read_online_analysis_data(None, 't0')
+                t0 = max(t0, int(data['t0'][0]))
+                logger.debug('t0 found; using initial t0 = {} instead of 1'.format(t0))
+            except Exception as e:
+                # No t0 found
+                logger.debug('Could not find t0: {}'.format(e))
+                pass
+
+            # Discard equilibration samples.
+            # TODO: if we include u_n[0] (the energy right after minimization) in the equilibration detection,
+            # TODO:         then number_equilibrated is 0. Find a better way than just discarding first frame.
+            i_t, g_i, n_effective_i = multistate.utils.get_equilibration_data_per_sample(u_n[t0:])
+            n_effective_max = n_effective_i.max()
+            i_max = n_effective_i.argmax()
+            n_equilibration = self._n_equilibration_iterations if self._n_equilibration_iterations is not None else i_t[i_max] + t0 # if self._n_equilibration_iterations was not specified, account for initially discarded frames
+            g_t = self._statistical_inefficiency if self._statistical_inefficiency is not None else g_i[i_max]
```

Comment:
I think we should use better names for the variables in this part, avoiding single letter names as possible. Just to improve readability when we have to come back to this code.

Reply:
I also see in this part that we use the specified values for `_statistical_inefficiency` and `_n_equilibration_iterations`, separately, but I wonder if specifying only one of them would need to affect the `n_effective_max` as in line 2028. I'm not really familiar with these calculations, but thought of bringing this up in case we need to find a more clever way to compute this values such that they use the specified parameters.

Reply:
All of the variable names here are the same as they were before, I didn't change them. `i` here represents index, so I could rename that to `index` if you think that's better. `g` represents the statistical inefficiency, and `t0` represents the number of equlibrations to discard due to equilibration. These variable choices are explained here: http://getyank.org/latest/algorithms.html#autocorrelate-algorithm

Reply:
> I wonder if specifying only one of them would need to affect the n_effective_max as in line 2028. 

Specifying only `_statistical_inefficiency` means that we will need to use `_get_equilibration_data_per_sample()` to compute the  `n_effective_max` because its used to compute `n_equilibration`.

Meanwhile, specifying only `n_equilibration` means that we need to use `_get_equilibration_data_per_sample()` to compute the  `n_effective_max` because its used to compute `g_t` (i.e. the statistical inefficiency).